src/main/resources/com/owlcyberdefense/syslog/xsd/syslog-cisco.dfdl.xsd

```diff
@@ -0,0 +1,136 @@
+    </xs:complexType>
+  </xs:element>
+
+  <xs:complexType name="Cisco">
+    <xs:annotation>
+      <xs:documentation>
+        Defines a type containing the content of a Cisco syslog message,
+        excluding the facility and severity. This should be used when the
+        faciliy/severity are already parsed.
+      </xs:documentation>
+    </xs:annotation>
+    <xs:sequence>
+      <xs:group ref="sl:Cisco" />
+    </xs:sequence>
+  </xs:complexType>
+
+  <xs:group name="Cisco">
+    <xs:annotation>
+      <xs:documentation>
+        Defines all content of a Cisco syslog message that appears immediately
+        affter the severify/facility part of a syslog message.
+      </xs:documentation>
+    </xs:annotation>
+    <xs:sequence>
+      <xs:element name="Header">
+        <xs:complexType>
+          <xs:sequence dfdl:separator="%SP;" dfdl:separatorPosition="postfix">
+            <xs:element name="SequenceNumber" type="xs:int" dfdl:textNumberPattern="0" dfdl:terminator=":" />
+            <xs:element name="TimestampMonth" dfdl:lengthKind="explicit" dfdl:length="3">
+              <xs:simpleType>
+                <xs:restriction base="xs:string">
+                  <xs:enumeration value="Jan" />
+                  <xs:enumeration value="Feb" />
+                  <xs:enumeration value="Mar" />
+                  <xs:enumeration value="Apr" />
+                  <xs:enumeration value="May" />
+                  <xs:enumeration value="Jun" />
+                  <xs:enumeration value="Jul" />
+                  <xs:enumeration value="Aug" />
+                  <xs:enumeration value="Sep" />
+                  <xs:enumeration value="Oct" />
+                  <xs:enumeration value="Nov" />
+                  <xs:enumeration value="Dec" />
+                </xs:restriction>
+              </xs:simpleType>
+            </xs:element>
+            <xs:element name="TimestampDay" dfdl:lengthKind="explicit" dfdl:length="2" dfdl:textNumberPattern="* 00">
+              <xs:simpleType>
+                <xs:restriction base="xs:int">
+                  <xs:minInclusive value="1" />
+                  <xs:maxInclusive value="31" />
+                </xs:restriction>
+              </xs:simpleType>
+            </xs:element>
+            <xs:element name="TimestampTime" type="xs:time" dfdl:calendarPattern="HH:mm:ss" dfdl:lengthKind="explicit" dfdl:length="8" dfdl:terminator=":" />
+            <xs:sequence dfdl:initiator="%%" dfdl:separator="-" dfdl:terminator=":">
+              <xs:element name="Facility" type="xs:string" />
+              <xs:element name="Severity" dfdl:textNumberPattern="0">
+                <xs:simpleType>
+                  <xs:restriction base="xs:int">
+                    <xs:minInclusive value="0" />
+                    <xs:maxInclusive value="7" />
+                  </xs:restriction>
+                </xs:simpleType>
+              </xs:element>
+              <xs:element name="Mnemonic" type="xs:string" />
```

Comment:
Grep everything for xs:string and replace all usages with your own simpleType with some maxLength facet.

Reply:
The spec doesn't specify max lengths for fields, and different fields would probably have very different lengths. For example, the Cisco Facility field is a short word like "PARSER", but the Message field is going to be significantly longer since it's the actual log message. We could pick a large maxLength of strings like 2048 to make XSAT happy and ensure all log messages pass except those that are clearly excessive, but different environments might know of, and want, tighter restrictions. And for fields like "Facility" that's waay to big of a max length. There's only 20 strings, so doing them individually probably wouldn't be too bad, but I'd basically just be guessing what numbers are reasonable, and could easily be wrong.

Reply:
Actually, RFC5424 does provide maxLengts for some header fields, which we do use. Other formats have similar fields so we can carry over maxLengths for those. The only one that doesn't have a max length is the message, but a large value like 9999 seems reasonable.